stan/math/prim/prob/wiener4_lcdf.hpp

```diff
@@ -0,0 +1,762 @@
+      prob = log1p(-exp_arg) - log_diff_exp(minus_two_va_one_minus_w, two_vaw);
+    } else {
+      prob = log1p(-exp_arg) - NEGATIVE_INFTY;
+    }
+  } else {
+    const auto exp_arg = exp(-minus_two_va_one_minus_w);
+    if (exp_arg >= nearly_one)
+      return ret_t(log1p(-w));
+    prob = log1p(-exp_arg) - log1p(-exp(2 * v * a));
+  }
+  return prob;
+}
+
+/**
+ * Calculate the probability term 'P' on log scale for grad_a and grad_v
+ *
+ * @param a The boundary separation
+ * @param v The drift rate
+ * @param w The relative starting point
+ * @return 'P' term
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto log_probability_GradAV(const T_a& a, const T_v& v,
+                                   const T_w& w) noexcept {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  auto nearly_one = ret_t(1.0 - 1.0e-6);
+  if (fabs(v) == 0.0) {
+    return ret_t(-w);
+  }
+  nearly_one = ret_t(1.0 - 1.1 * 1.0e-5);
+  ret_t prob;
+  if (v < 0) {
+    const auto two_va_one_minus_w = (2.0 * v * a * (1.0 - w));
+    const auto two_avw = 2 * a * v * w;
+    const auto two_av = 2 * a * v;
+    const auto exp_two_va_one_minus_w = exp(two_va_one_minus_w);
+    const auto exp_two_avw = exp(two_avw);
+    const auto exp_two_av = exp(two_av);
+    if (((exp_two_va_one_minus_w >= nearly_one) || (exp_two_avw >= nearly_one))
+        || (exp_two_av >= nearly_one)) {
+      return ret_t(-w);
+    }
+    prob = LOG_TWO + two_va_one_minus_w - log1p(-exp_two_va_one_minus_w);
+    auto log_quotient = log1p(-exp_two_avw) - log1p(-exp_two_av);
+    if (log(w) > log_quotient) {
+      prob += log_diff_exp(log(w), log_quotient);
+      prob = exp(prob);
+    } else {
+      prob += log_diff_exp(log_quotient, log(w));
+      prob = -exp(prob);
+    }
+  } else {
+    const auto minus_two_va_one_minus_w = (-2.0 * v * a * (1.0 - w));
+    const auto minus_two_av = (-2 * a * v);
+    const auto exp_minus_two_va_one_minus_w = exp(minus_two_va_one_minus_w);
+    const auto exp_minus_two_av = exp(minus_two_av);
+    if ((exp_minus_two_va_one_minus_w >= nearly_one)
+        || (exp_minus_two_av >= nearly_one)) {
+      return ret_t(-w);
+    }
+    prob = LOG_TWO - log1p(-exp_minus_two_va_one_minus_w);
+    ret_t log_quotient;
+    if (minus_two_va_one_minus_w > minus_two_av) {
+      log_quotient = log_diff_exp(minus_two_va_one_minus_w, minus_two_av)
+                     - log1p(-exp_minus_two_av);
+    } else if (minus_two_va_one_minus_w < minus_two_av) {
+      log_quotient = log_diff_exp(minus_two_av, minus_two_va_one_minus_w)
+                     - log1p(-exp_minus_two_av);
+    } else {
+      log_quotient = NEGATIVE_INFTY - log1p(-exp_minus_two_av);
+    }
+    if (log(w) > log_quotient) {
+      prob += log_diff_exp(log(w), log_quotient);
+      prob = -exp(prob);
+    } else {
+      prob += log_diff_exp(log_quotient, log(w));
+      prob = exp(prob);
+    }
+  }
+  return (is_inf(prob)) ? NEGATIVE_INFTY : prob;
```

Comment:
Inf is fine no need for the check here

Reply:
Ok. Deleted and changed to early return.